### PR TITLE
DataViews Extensibility: Add a hook to allow third-party scripts to register/unregister post type actions

### PR DESCRIPTION
**Pull request by youknowriad**
Related #61084 

## What?

In #62052 an API to register and unregister dataviews actions has been implemented. But in order to allow third-party developers to be able to unregister these actions properly we need to offer a hook that third-party can use to "time" their register/unregister function calls properly. This PR does so by introducing the `core.registerPostTypeActions` hook to do so.

So something like:

```
wp.hooks.addAction(
	'core.registerPostTypeActions',
	'my-plugin/register-post-type-actions',
	( postType ) => {
            if ( postType === 'post' ) {
                wp.editor.registerEntityAction( 'postType', 'post', myCustomAction );
                wp.editor.unregisterEntityAction( 'postType', 'post', 'permanently-delete' );
            }
	}
);
```

Another goal of this PR is to allow us (core) to use the "post type" object to register/unregister actions. For instance "viewPost" action need to only be register for post types with is_viewable as true...

## Testing Instructions

Check that the post actions dropdown work properly in both the dataviews and the editor.

### CODE REVIEW

**Comment by youknowriad** [comment]
One small (or not) performance optimization in this PR is that the actions are only registered when needed, they are not globally registered anymore.

**Comment by gziolo** [comment]
Overall, it’s a common problem that we have with plugins which often run before the editor registers its default handlers, the same applies to block types, block variations. I was thinking again about replicating:

```php
add_action( ‘init’, ‘someCallback’ );
```

Today, we offer in the documentation a workaround with

```php
wp.domReady( someCallback );
```

Example: [removing registered block types](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/filters/block-filters.md#using-a-deny-list).

However, it’d definitely help to have it more formalized. It’s only the question whether we introduce specific actions like in this PR, mix that with some general purpose:

```js
wp.hooks.addAction( ‘editor/init’, ‘someCallback’ );
```
Or we offer both. 

Concluding, I’m very much in favor of the proposed approach. I’m also happy to discuss further the overall strategy for other extensibility points that addresses the developer experience hurdles when plugins run code to early by introducing familiar patterns as they have to use in PHP that ensure the code runs at the right time.

**Comment by youknowriad** [comment]
Pros here:

 - Performance
 - Not impacted by script loading...

Cons:

 - Lose some reactivity, harder to unregister/register as some dependency change...

I think it's the right approach though, we've been accumulating these global registration functions and the limit is showing.

**Comment by gziolo** [comment]
> Lose some reactivity, harder to unregister/register as some dependency change...

Can you share more details regarding that aspect? Do you mean the problem that some plugins can run after the `doAction( 'editor/init' )` call? 

Can we avoid it by implementing a similar solution to [`withFilters` HOC](https://github.com/WordPress/gutenberg/blob/3621ab3b8ea556a4af367ba2b5e91e748b36791e/packages/components/src/higher-order/with-filters/index.tsx#L46) which retriggers this action whenever a new action gets registered? 

One good aspect would be that when `editor/init` fires, all plugin authors can assume that at least the core finished registering everything that is essential.

File: packages/editor/src/components/post-actions/actions.js [hunk 1]
```diff
@@ -589,6 +589,10 @@ export function usePostActions( { postType, onActionPerformed, context } ) {
 		},
 		[ postType ]
 	);
+	const { registerPostTypeActions } = unlock( useDispatch( editorStore ) );
+	useEffect( () => {
+		registerPostTypeActions( postType );
+	}, [ registerPostTypeActions, postType ] );
```
Comment (ellatrix) [inline packages/editor/src/components/post-actions/actions.js:595]: I don't love these kinds of effects tbh. Not sure what the alternative is, I guess filtered context wouldn't work.

File: packages/editor/src/dataviews/store/private-actions.ts [hunk 2]
```diff
@@ -28,3 +43,56 @@ export function unregisterEntityAction(
...
+				}
+				unlock( registry.dispatch( editorStore ) ).registerEntityAction(
+					'postType',
+					postType,
+					action
+				);
+			} );
+		} );
+
+		doAction( 'core.registerPostTypeActions', Promise.resolve(), postType );
```
Comment (ellatrix) [inline packages/editor/src/dataviews/store/private-actions.ts]: Could our action registration use the same hook?

**Review by ellatrix — APPROVED** [review]
What are the changes to isReady for?

**Comment by youknowriad** [comment]
The `isReady` is to avoid triggering the action twice, (say you open the dataviews actions, navigate away and come back)

File: packages/editor/src/dataviews/store/private-actions.ts (hunk 2 above)
Comment (youknowriad) [reply packages/editor/src/dataviews/store/private-actions.ts]: I can see two sides here:

 - Core code is not special and should use the same thing.
 - Adds complexity: figuring out when to register the action, which package and also no special treatment for performance...

File: packages/editor/src/components/post-actions/actions.js (hunk 1 above)
Comment (youknowriad) [reply packages/editor/src/components/post-actions/actions.js:595]: This is for me a good thing: Only register the actions and run the filter when needed (when using usePostAction hook), that said, I agree that the code is weird, but I don't know of another pattern that provides the same benefits.

**Comment by gziolo** [comment]
> Pros here:
> 
> * Performance
> * Not impacted by script loading...
> 
> Cons:
> 
> * Lose some reactivity, harder to unregister/register as some dependency change...
> 
> I think it's the right approach though, we've been accumulating these global registration functions and the limit is showing.

I drafted PR to keep the discussion going:
- https://github.com/WordPress/gutenberg/pull/65921

**Pull request merged**